### ISSUE TITLE
create mm module that changes variables with ! to nested variables

### ISSUE THREAD

**Issue by davidelang** [rsyslog/rsyslog#1281]
currently it is possible to create variables that can't be accessed because json input can create variables that have ! in their name and rsyslog treats ! as a way to separate nested variables. This can also be done with mmnormalize.

so you can have json arrive that is {"foo!bar":"baz} and get parsed, but if you look for $!foo!bar it will be looking for {"foo" :{"bar":"baz"}} instead.

this module should go through all the variable names looking for '!' and if it finds anything, change it from a single name to a nested structure (creating a new one or using an existing one for the upper levels)

**Comment by rgerhards** [comment]
good catch. Might there be a valid reason to have "!" inside the name (e.g. like DOT in ES? ;-)). I think about cases where we get data from extern and need to forward it to extern. If so, we should also consider some kind of escape mechanism.

**Comment by davidelang** [comment]
There's always a valid reason for any odd thing :-)

Another reason for having this module is that liblognorm doesn't let you define anything other than a top level item (except with user defined types), so if you want to parse something into $!source!ip you don't have a clean way to do it. You can define %source!ip:ipv4% and liblognorm will handle it, but create a variable that you can't currently access in rsyslog. This would fix that

I suspect that the times where you need to both access a variable that's got a ! in it, and forward it unmodified are very rare.

Remember that the ! character was picked for the separator was specifically because it was so unlikely to appear in a variable name (I'm blanking now on the now-dead log standardization project where that was discussed)

Adding the ability to escape characters in a rsyslog variable name is an interesting thought. We need to make a formal definition of what characters rsyslog allows in a name (json allows anything other than newline IIRC)

**Comment by rgerhards** [comment]
This is implemented by the `fmunflatten` function module. `unflatten(source, separator)` expands flat JSON keys into nested JSON; using `"!"` as the separator handles the exact `foo!bar` case described here. The implementation also has focused coverage in `tests/rscript_unflatten_object_exclamation.sh`.

Documentation: https://www.rsyslog.com/doc/rainerscript/functions/mo-unflatten.html
Source: https://github.com/rsyslog/rsyslog/tree/main/contrib/fmunflatten

The function was added in commit `fad41a617` (2020-08-16). Closing as implemented.
